### ISSUE TITLE
When used with get_mock_object, DateTimeType returns naive datetime

### ISSUE THREAD

**Issue by charlax** [schematics/schematics#388]
Fairly straightforward to reproduce:

```
In [1]: from schematics.types import DateTimeType

In [2]: from schematics.models import Model

In [11]: class Toaster(Model):
    created_at = DateTimeType(required=True)

In [15]: Toaster.get_mock_object().created_at.tzinfo

In [16]:
```

Since people might have assumed this would always be the case, perhaps an extra constructor arguments `default_timezone` would be useful? This would make sure that only timezone-aware datetime are returned. It would use the timezone on the parsed datetime if it's there, or the default one.

Thoughts?

**Comment by bintoro** [comment]
Okay, so the rewritten `DateTimeType` is very configurable but not to the extent that you could specify a "default" timezone other than UTC.

When rewriting the class I have totally neglected `_mock()`, which means that when the field is configured to allow aware datetimes

**Comment by bintoro** [comment]
Okay, so the rewritten DateTimeType is very configurable but not to the extent that you could specify a "default" timezone other than UTC.

When rewriting the class I have totally neglected `_mock()`, which means that even if the field is configured to allow aware datetimes *only*, `_mock()` still yields a naive datetime. This is clearly a bug.

By default both naive and aware datetimes are accepted. Perhaps `_mock()` should then randomly return one or the other.

**Comment by charlax** [comment]
What about backward compatibility? Since currently `_mock()` returns naive datetime, some people's tests will start breaking if it returns aware/naive datetime at random... This would be a breaking change.

I'd much prefer something explicit where naive datetime are returned by default, unless `timezone=pytz.utc` (for instance) is used. Or we can have a dedicated `DateTimeWithTimezone` object but that seems a bit heavy-handed.

**Comment by bintoro** [comment]
Rewriting the type to accept inputs both with and without a time zone designator by default was a kind of breaking change in itself. I'd say it's correct to change `_mock()` to reflect this and break people's tests rather than (potentially) their running applications.

There was no good way to preserve full backward compatibility because the old behavior was so bizarre. The field would accept a `Z` suffix but silently discard it; the outcome would always be a naive datetime. It's hard to reconcile that with any form of timezone support that makes sense.

(#310 is the patch that introduced the `Z` suffix in the input pattern. It's still relatively recent, so we *could* just pretend it never existed.)

For more context, see the discussion in #358.

The trouble with datetimes is that there's no universal setup that does the right thing out of the box. `DateTimeType` is a field people really should configure for themselves. A major goal of the rewrite was to make that configuration possible.

I would gladly welcome any suggestions to improve the new design before v2.0 is released.

**Comment by charlax** [comment]
The problem with the random behavior is that an exception will be raised when comparing a naive datetime with an aware one. Yet if on production I always expect the datetime to be timezone-aware, I see no reason why mock object created by schematics include naive timezone. This will force people to either change the `_mock` to always return timezone-aware datetime, or change their code to accommodate the tests (which is a big no-no).

So to recap the options:

1. Always return aware and naive datetime at random.
2. Observe `tzd` parameter to return the correct mocked value.

Logic for option 2.:

* `tzd == 'reject'`: always return naive datetime.
* `tzd == 'allow'`: return naive/aware datetime at random.
* `tzd == 'require'`: return timezone-aware datetime in random timezone.
* `tzd == 'utc'`: return timezone-aware datetime in UTC.

This seems to be pretty intuitive.

cc @mjrk and @j2labs.

**Comment by bintoro** [comment]
@charlax "Option 2" is precisely what I had in mind.

I wrote:

> By default both naive and aware datetimes are accepted. Perhaps `_mock()` should then randomly return one or the other.

That is, randomly return one or the other **by default**, i.e., when `tzd == 'allow'`.

**Comment by charlax** [comment]
Yep - was reading the implementation on `master` instead of `development` for my first comment.

So if I was to implement 2 in a PR it'd be accepted?

**Comment by bintoro** [comment]
Yes :)

There's `DateTimeType.offset_timezone` providing an implementation for fixed-offset time zones.

**Comment by bintoro** [comment]
Fixed in 2110749.
